### DISCUSSION TITLE
How to add custom state on AsyncValue?

### DISCUSSION THREAD

**Discussion by td026544** [rrousselGit/riverpod#985] [Q&A]
I want to add custom state (AsyncEmpty) in AsyncValue.
 
Can anyone help?

**Comment by rrousselGit** [comment]
It's discouraged.

Instead consider doing `AsyncValue<MaybeEmpty<T>>`

**Reply by japscodelabs** [reply]
Can I know the reason for this? I've been using bloc pattern and it was a practice to put state in the following
- Initial
- Loading
- Error
- Data

while async value has the only state

**Reply by rrousselGit** [reply]
Because that's not something easily doable, and Riverpod relies on knowing whether a provider uses AsyncValue or not

**Reply by japscodelabs** [reply]
What do you mean by this? Because that's not something easily doable.

**Reply by iamarnas** [reply]
@japscodelabs If you combine `AsyncValue` with `StateNotifier` you can put initial data directly to `AsyncData` and control loading state how you want on loading new data.

**Reply by japscodelabs** [reply]
Im just curious if it is a save button/action that uses async value
does that mean the initial value is AsyncValue.data(null)?

Because the loading value = will be a disabled button with circular indicator
error value = will be a retry labeled button
data value = will probably a success indicator button like for example green button with check icon.

Will you recommended creating own asyncvalue like implementation with initial?
or it is recommended to use AsyncValue.data(null) as initial state?

More accurately if im using Option FP does it should be AsyncValue.data(Option.none());

**Reply by iamarnas** [reply]
> Im just curious if it is a save button/action that uses async value
> does that mean the initial value is AsyncValue.data(null)?
> 
> Because the loading value = will be a disabled button with circular indicator
> error value = will be a retry labeled button
> data value = will probably a success indicator button like for example green button with check icon.
> 
> Will you recommended creating own asyncvalue like implementation with initial?
> or it is recommended to use AsyncValue.data(null) as initial state?
> 
> More accurately if im using Option FP does it should be AsyncValue.data(Option.none());

If you put to StateNotifier super constructor AsyncValue.data(0) then will be not null. Then inside the the class you can create any functions and control state how you want.

**Reply by rrousselGit** [reply]
> Will you recommended creating own asyncvalue like implementation with initial?

As I said before, don't. Riverpod relies on knowing that a Provider uses AsyncValue
If you're making your own, Riverpod won't know about it.


AsyncValue.data(null) or AsyncValue.data(Option.none()) are both fine.
